units/event_metajob.go

```diff
@@ -159,28 +162,51 @@ func (j *eventMetaJob) dispatchLoop(ctx context.Context) error {
 	startTime := time.Now()
 	logger := event.NewDBEventLogger(event.AllLogCollection)
 	catcher := grip.NewSimpleCatcher()
-	notifications := make([][]notification.Notification, len(j.events))
 
-	var err error
-	for i := range j.events {
-		notifications[i], err = tryProcessOneEvent(&j.events[i])
-		catcher.Add(err)
-		if err = notification.InsertMany(notifications[i]...); err != nil {
-			grip.Error(message.WrapError(err, message.Fields{
-				"job_id":        j.ID(),
-				"job":           eventMetaJobName,
-				"source":        "events-processing",
-				"notifications": notifications[i],
-				"message":       "can't insert notifications",
-			}))
-			catcher.Add(err)
+	input := make(chan event.EventLogEntry, len(j.events))
+	for _, e := range j.events {
+		select {
+		case input <- e:
+			continue
+		case <-ctx.Done():
+			return errors.New("operation aborted")
 		}
-	}
 
-	for idx := range notifications {
-		catcher.Add(j.dispatch(notifications[idx]))
-		catcher.Add(logger.MarkProcessed(&j.events[idx]))
 	}
+	close(input)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer recovery.LogStackTraceAndContinue("problem during notification dispatch")
+
+			for e := range input {
+				n, err := tryProcessOneEvent(&e)
+				catcher.Add(err)
+				if err != nil {
+					continue
+				}
+
+				if err = notification.InsertMany(n...); err != nil {
+					grip.Error(message.WrapError(err, message.Fields{
+						"job_id":        j.ID(),
+						"job":           eventMetaJobName,
+						"source":        "events-processing",
+						"notifications": n,
+						"message":       "can't insert notifications",
+					}))
+					catcher.Add(err)
+					continue
```

Comment:
To Brian's point: notification _ids are intended to collide (which is why the insert is unordered). If we don't dispatch when we hit an error on the insert, none of the notifications from this event will be processed.